_posts/2024-07-03-#29775.md

```diff
@@ -4,24 +4,41 @@ date: 2024-07-03
+    - Testnet reset [discussion on the mailing list](https://groups.google.com/g/bitcoindev/c/9bL00vRj7OU).
+    - A helpful [deep dive into the block storm bug by Jameson Lopp](https://blog.lopp.net/the-block-storms-of-bitcoins-testnet/).
+    - A great [explanation of the timewarp attack on Bitcoin SE by David Harding](https://bitcoin.stackexchange.com/questions/75831/what-is-time-warp-attack-and-how-does-it-work-in-general/75834#75834)
+- Bitcoin core supports three test networks out of the box: Regtest, Testnet, and Signet. There also exist some custom Signet variants like Mutinynet. At this point, the current Testnet has been running for 12 years. Pepperidge Farm remembers that this is not the original Testnet though, it is actually Testnet 3. It was introduced in [PR #1392](https://github.com/bitcoin/bitcoin/pull/1392). Documentation on how exactly Testnet 1 and 2 broke is not available but it appears that they fell victim to high fluctuation in mining power. Remember that around this time the first ASIC miners entered the market while Testnet was probably still mostly mined by CPUs and maybe the occasional GPU.
+- Testnet 3 features a Proof of Work exception rule, known as the 20-min exception. This rule was designed to prevent the chain from getting stuck again due to hash power fluctuation. However, a bug in this exception leads to so-called block storms, large numbers of blocks being mined in quick succession. This is the main reason Testnet 3 is so far ahead of mainnet even though it started much later. The bug was recently [exploited on purpose](https://www.ccn.com/news/crypto/bitcoin-testnet-could-need-reset-13-years-jameson-lopp-griefing/) for an extended period of time to highlight the issue.
+- Testnet 4 still includes the 20-min exception but adds a mitigation for the block storm issue.
+- The pull request also includes a fix for the timewarp attack, an attack that is still possible on mainnet today. A fix for this was proposed as part of the [Great Consensus Cleanup](https://bitcoinops.org/en/topics/consensus-cleanup-soft-fork/) but failed to get the necessary support as a softfork so far.
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
-
+2. What is the message in the Genesis block in Testnet 3 and why (reference the code)?
+3. Aside from the consensus changes, what differences do you see between Testnet 4 and Testnet 3, particularly the chain params?
+4. Pick a single chain param that you don't know/remember the meaning of. Look up what it does and explain it in one sentence.
+5. How does the 20-min exception rule work in Testnet 3? How does this lead to the block storm bug? Please try to reference the code.
+6. How is the block storm bug fixed in the PR? What other fixes were discussed in the PR?
+7. Why was the time warp fix included in the PR?
```

Comment:
Might be helpful to link to a place where you stated why

Reply:
I assumed you didn't want me to link directly to the comment since that would make it too easy :) I added a vague hint.